src/modules/ModuleLinker.v3

```diff
@@ -0,0 +1,233 @@
+// Copyright 2023 Wizard Authors. All rights reserved.
+// See LICENSE for details of Apache 2.0 license.
+
+/// Links the main module with the pre-loaded modules, and returns the main module instance.
+///
+/// The linker partially instantiates all the modules, then resolve each import to its original
+/// provider, then completes the instantiation. This allows bi-directional imports among multiple modules,
+/// as long as the import chain is acyclic.
+class ModuleLinker(engine: Engine,
+				   program_args: Array<string>,
+				   names: Array<string>,
```

Comment:
It'd be a little more robust if these two arguments were an array of tuples, rather than parallel arrays of the same length.